### PR TITLE
Feature/quick play to join a minecraft server, world and realm right after launching the game

### PR DESCRIPTION
**Pull request by ellet0**
### Go to the bottom of this PR

This will allow the player to configure the instance so that when he or she plays the game by launching the instance, it will automatically join into a Minecraft server that is specified by the user

image

This feature is designed for newer Minecraft versions ([23w14a](https://www.minecraft.net/en-us/article/minecraft-snapshot-23w14a) and up) but fully compatible with `23w13a` and older versions

image

image

Feel free to ask for any changes to maintain consistency, follow the project guidelines, and make better naming or any additional changes to get the best code quality possible

The art of naming in Programming:

Should we call it `Join server after launch` instead? since it's launching the game first and then joining a server, but the current name might be preferred by users

What do you want the property to be called?

- `joinInitialServerAddress`
- `initialJoinServerAddress`
- `initialServerJoinAddress`
- `joinServerOnLaunchAddress`
- `joinServerAddress`
- `initialJoinServerAddress`
- `joinServerInitialAddress`
- `onLaunchJoinServerAddress`
- `joinServerAddress`

The list can be much larger. suggest the best name possible that fit with the variable/property naming in this project
because changing it later would require breaking changes or deprecation/migration 

This feature is inspired by another launcher that is based on MultiMC

Suggestion: When importing an instance from another launcher format, we might want to add support for this feature so if the launcher has this feature, it will import the `server address to join` property into the ATLauncher instance

Out of scope for this PR but can be used:

This will also allow Minecraft scripts like [this](https://github.com/ellet0/minecraft-sync) to update the server address property in `instance.json`so it will always be synced with the latest server

### Testing

I tested this feature with the following versions:

- `1.20.6` with and without Fabric
- `1.20.1` with and without Fabric
- `1.19.4`
- `23w13a`
- `23w14a`

I made it so that when you enter an address without a port it will use `25565` as default just like Minecraft would do it, or if you enter the full address with the port, it will also work as expected

This PR doesn't have any tests in the source code because I can't find existing tests related to the changes 

### Related Issues

- #748

### Quick Play Feature PR Update

This PR was initially created to support joining a server on the game as it's supported in the game args on both older and newer versions of Minecraft, plans have changed, I could create a new PR but then this should be approved first as the new one based on it, don't get confused by the branch name in my fork

image

In addition to everything above, I have updated this PR to include support for single-player and realm (currently work only on 23w14a and above so I left a note in the help icon but also left a TODO for adding backward compatibility so anyone can add support if needed)

I added a validation for all three inputs in to handle all the possible cases (like when the user doesn't have any world and chooses single player then tries to save the changes)

I tested the changes more, for example when selecting a world deleting the save world, and then trying to launch the game, the logic will work in a way that it won't throw an exception and instead will select a different world but needs to save the world once you select a world in the launcher, delete in the instance files, then you will get an error in the game and to fix it the player needs to go to the settings of the instance once again, select a different world (will be handled by default) then click save. this whole change in the UI would have been a lot easier if we used declarative UI like Compose desktop with Kotlin but for the current solution, we should improve the UI and separate some of the logic, extract a lot of common code, improve the naming and other important improvements before merging this PR

### CODE REVIEW

**Comment by RyanTheAllmighty** [comment]
Nice, we have an issue up for this https://github.com/ATLauncher/ATLauncher/issues/748

Ideally it'd also be good to support single player worlds (with a dropdown of all world paths in the filesystem, validated on launch so that if it's deleted it will not use the path when launching) as well as realms (which requires passing a realms id, since we have the users auth tokens we might be able to call some api to get a list of realms, but sounds a bit non trivial to check and keep that up to date).

I'll leave it up to you if you want to investigate adding in singleplayer and realm support, completely happy to leave it out of this PR.

For the naming, I think what you have on the UI is fine, as for any underlying variable naming, I'd be for calling it more closely to what Minecraft calls the feature, in this case `quickPlayMultiplayer`, `quickPlaySingleplayer` and `quickPlayRealms` as is defined in the minecraft json files.

But yeah as per your comments on testing, don't worry about it, we do minimal testing, it's been on the todo list for like 8 years to improve.

Thanks for contributing, please also feel free to join the [DIscord](https://atl.pw/discord) and use the #development channel if you want to chat about any issues, launcher improvements, prs etc.

File: src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java
Comment (RyanTheAllmighty) [inline src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:159]: We'll want to do some validation here to make sure it's valid, otherwise any old garbage could be entered here.

I guess something simple like check that there's either no `:` or if there is a `:` that there are only numbers after the `:` between 1 and 65535. That should be simple enough without having to get into doing 500 character long regexes.

**Review by RyanTheAllmighty — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java
Comment (ellet0) [reply src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:159]: > We'll want to do some validation here to make sure it's valid, otherwise any old garbage could be entered here.
> 
> I guess something simple like check that there's either no `:` or if there is a `:` that there are only numbers after the `:` between 1 and 65535. That should be simple enough without having to get into doing 500 character long regexes.

I wanted to discuss about that, but I didn't find any section for the suggestions and there is no discussions second in this repository and I don't plan on mixing the issues with the questions for organization, and an example of how ATLauncher handles the input validation so I do something similar here for consistency

File: src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java
Comment (ellet0) [reply src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:149]: > While I need to create some more generic ui components to wrap common logic like this, adding a x to the input like on search in packs might be nice, but completely optional:
> 
> ```suggestion
>         initialJoinServerAddress = new JTextField(13);
>         initialJoinServerAddress.putClientProperty("JTextField.showClearButton", true);
>         initialJoinServerAddress.putClientProperty("JTextField.clearCallback", (Runnable) () -> {
>             initialJoinServerAddress.setText("");
>         });
> ```

Sure

File: CHANGELOG.md
Comment (ellet0) [reply CHANGELOG.md:10]: > Linking to the issue and like `[#748]` instead of a full link here is good, we convert it to a full link during the release process
> ```suggestion
> - Add the option to join a minecraft server when launching a instance [#748]
> ```

I will update it

File: src/main/java/com/atlauncher/data/InstanceLauncher.java
Comment (ellet0) [reply src/main/java/com/atlauncher/data/InstanceLauncher.java:71]: > As mentioned I think using the Mojang defined variable name would be good here
> 
> ```suggestion
>     public String quickPlayMultiplayer = null;
> ```

But maybe we might want to be more specific? Adding (serverAddress) so developers that have scripts that deal with the json can understand it?

**Comment by ellet0** [comment]
> Ideally it'd also be good to support single player worlds (with a dropdown of all world paths in the filesystem, validated on launch so that if it's deleted it will not use the path when launching) as well as realms (which requires passing a realms id, since we have the users auth tokens we might be able to call some api to get a list of realms, but sounds a bit non trivial to check and keep that up to date).

> I'll leave it up to you if you want to investigate adding in singleplayer and realm support, completely happy to leave it out of this PR.

Sure, I would work on it after finishing this feature, but how it will look in the UI? Should we add a new dropdown input field that allow selecting a world and the user should either select a world or a server?

also as a bonus, we might add a text suggestion for the list of the current servers so the player can select a server directly but I have zero idea how to read `servers.dat` file or how to get the server list in the game

> For the naming, I think what you have on the UI is fine, as for any underlying variable naming, I'd be for calling it more closely to what Minecraft calls the feature, in this case quickPlayMultiplayer, quickPlaySingleplayer and quickPlayRealms as is defined in the minecraft json files. Keeping the UI elements named as they are (initialJoinServerAddress) is good so it ties to the name of the setting, just the actual underlying setting value I think would be good to name as mojang have

as for that, do you think we should be more specific so anyone who edit the data of instance in json file can understand it (`quickPlayerMultiplayer` might be not very clear) ? but instead of `quickPlayMultiplayerServerAddress` how about we create a new data class for the `quickPlay` and will have different options, one for multiplayer or server address, another for world name or singlePlayer etc...


> But yeah as per your comments on testing, don't worry about it, we do minimal testing, it's been on the todo list for like 8 years to improve.

Me too, I'm not tests expert but I can write tests for most things I do but I only write a very little.

> Please also feel free to join the [DIscord](https://atl.pw/discord) and use the #development channel if you want to chat about any issues, launcher improvements, prs etc.Thanks for contributing, please also feel free to join the [DIscord](https://atl.pw/discord) and use the #development channel if you want to chat about any issues, launcher improvements, prs etc.****

Sure, I have a few questions about the project

**Comment by ellet0** [comment]
One more thing, if we are going to add support for a quick single-player feature too, then maybe we should have a category or sub-category in the `General` tab. maybe moving the text input field to a different tab (General, Java/Minecraft) or creating its own tab (Quick multiplayer or running options or Misc tab?)

File: src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java
Comment (ellet0) [reply src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:163]: > I guess something simple like check that there's either no `:` or if there is a `:` that there are only numbers after the `:` between 1 and 65535. That should be simple enough without having to get into doing 500 character long regexes.

Should the Patterns or Input text validators go to the source code? I want to add some tests for validating input and output, regarding of using Pattern or validating the input using java internal functions like (contains, isEmpty etc..) it could be a function and not directly a Pattern because we might change the Pattern later and have to change it from different places, also tests are helpful to make sure we are safe

File: src/main/java/com/atlauncher/data/InstanceLauncher.java
Comment (RyanTheAllmighty) [reply src/main/java/com/atlauncher/data/InstanceLauncher.java:71]: My thinking was that Mojang named it as `quickPlayMultiplayer` so matching it up with Mojangs own naming would be good. But also happy if you want to call it `quickPlayMultiplayerServerAddress` or alternatively turn it into an object for all 3 of the quick play settings.

Just best to decide now as changing later is not trivial once out in public

**Comment by RyanTheAllmighty** [comment]
> > Ideally it'd also be good to support single player worlds (with a dropdown of all world paths in the filesystem, validated on launch so that if it's deleted it will not use the path when launching) as well as realms (which requires passing a realms id, since we have the users auth tokens we might be able to call some api to get a list of realms, but sounds a bit non trivial to check and keep that up to date).
> 
> > I'll leave it up to you if you want to investigate adding in singleplayer and realm support, completely happy to leave it out of this PR.
> 
> Sure, I would work on it after finishing this feature, but how it will look in the UI? Should we add a new dropdown input field that allow selecting a world and the user should either select a world or a server?
>
> also as a bonus, we might add a text suggestion for the list of the current servers so the player can select a server directly but I have zero idea how to read `servers.dat` file or how to get the server list in the game

Hmm good point, we would only want to allow one of the quick play options. Only thing I can think of right now is a dropdown box to select a server, world or realm, and then a single input that changes based on which is selected. Or alternatively a radio button group might fit better

But yeah reading the servers.dat I've wanted to do for a while, not sure how difficult it is, but I imagine it's something fairly simple. But also nice to have rather than necessary right now.

> > For the naming, I think what you have on the UI is fine, as for any underlying variable naming, I'd be for calling it more closely to what Minecraft calls the feature, in this case quickPlayMultiplayer, quickPlaySingleplayer and quickPlayRealms as is defined in the minecraft json files. Keeping the UI elements named as they are (initialJoinServerAddress) is good so it ties to the name of the setting, just the actual underlying setting value I think would be good to name as mojang have
> 
> as for that, do you think we should be more specific so anyone who edit the data of instance in json file can understand it (`quickPlayerMultiplayer` might be not very clear) ? but instead of `quickPlayMultiplayerServerAddress` how about we create a new data class for the `quickPlay` and will have different options, one for multiplayer or server address, another for world name or singlePlayer etc...

Happy with either as mentioned in my other comment

> > But yeah as per your comments on testing, don't worry about it, we do minimal testing, it's been on the todo list for like 8 years to improve.
> 
> Me too, I'm not tests expert but I can write tests for most things I do but I only write a very little.
> 
> > Please also feel free to join the [DIscord](https://atl.pw/discord) and use the #development channel if you want to chat about any issues, launcher improvements, prs etc.Thanks for contributing, please also feel free to join the [DIscord](https://atl.pw/discord) and use the #development channel if you want to chat about any issues, launcher improvements, prs etc.****
> 
> Sure, I have a few questions about the project

File: src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java
Comment (ellet0) [reply src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:163]: > Would make sense to have a utility class for validation, like ValidationUtils for instance (or simply just Validation)
Where should this File go? Maybe (src/main/java/com/atlauncher/utils)?? I'm not sure if there is a better package

Comment (ellet0) [reply src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:163]: I'm getting build failure exception:

```console

* What went wrong:
Execution failed for task ':checkLicenseMain'.
> License violations were found: src/main/java/com/atlauncher/utils/ValidationUtils.java

```

it's a simple class that doesn't use any imports other than `java.util.regex.Pattern`

Comment (ellet0) [reply src/main/java/com/atlauncher/gui/dialogs/instancesettings/GeneralInstanceSettingsTab.java:180]: > We'll want to do some validation here to make sure it's valid

Done, there are common logic in the validation and the error dialogs, maybe we should extract them later

**Comment by ellet0** [comment]
Any idea how to add a dropdown to a `JTextField` in Java Swing to allow the player to select a list of the servers as predefined values? so the player doesn't have to enter the server address manually if there are some servers in the game

I'm not a Java Swing expert but I would know how to do it in more modern frameworks like Compose, React or even Android XML Views etc..

File: src/main/java/com/atlauncher/data/InstanceLauncher.java
Comment (ellet0) [reply src/main/java/com/atlauncher/data/InstanceLauncher.java]: > or alternatively turn it into an object for all 3 of the quick play settings.
This has been solved and changed, it's time to review the code

**Comment by ellet0** [comment]
Should we add an option/checkbox to toggle disabling and enabling the quick play? We also could add other features like allow the player to select a world from the play dropdown (the one that has play and play offline)
So the player can customize which world, server, and realm to load without changing it every time in the settings of the instance, I'm not sure if this feature should be more like the original launcher or similar to it, I'm asking for your suggestions before merging and release the new version.

File: src/main/java/com/atlauncher/data/json/QuickPlay.java
Comment (ellet0) [inline src/main/java/com/atlauncher/data/json/QuickPlay.java:40]: Should we rename to `worldSaveName` to be more specific?

Comment (RyanTheAllmighty) [reply src/main/java/com/atlauncher/data/json/QuickPlay.java:40]: I think `worldName` is fine, names within the json are not that important as they're not meant to be edited manually

Comment (ellet0) [reply src/main/java/com/atlauncher/data/json/QuickPlay.java:40]: > I think `worldName` is fine, names within the json are not that important as they're not meant to be edited manually

The code is not fully documented so we should consider a descriptive name, it's not just for the JSON file, it's also in the data class in the source code so the new developers know what they are dealing with so will not be something that is not intended to avoid unnecessary bugs in the future.

File: src/main/java/com/atlauncher/mclauncher/MCLauncher.java
Comment (RyanTheAllmighty) [inline src/main/java/com/atlauncher/mclauncher/MCLauncher.java:448]: Now the issue with this and realms is there is no way to do this on older versions, so we can remove the empty else blocks here

It might be good to do these checks in the settings page and disable the singleplayer and realms options in the dropdown if it's not supported so they cannot be selected.

Comment (ellet0) [reply src/main/java/com/atlauncher/mclauncher/MCLauncher.java:448]: > Now the issue with this and realms is there is no way to do this on older versions, so we can remove the empty else blocks here
> 
> It might be good to do these checks in the settings page and disable the singleplayer and realms options in the dropdown if it's not supported so they cannot be selected.

Yes, I thought about that, but then we will end up having only one option for newer versions. Should we do the check using the game args or the game version? We might need a way to compare the versions correctly

Comment (RyanTheAllmighty) [reply src/main/java/com/atlauncher/mclauncher/MCLauncher.java:448]: Game args check like you've done here should be fine since it's only supported on versions that have these params anyway

File: src/main/java/com/atlauncher/data/json/QuickPlay.java
Comment (ellet0) [reply src/main/java/com/atlauncher/data/json/QuickPlay.java]: Updated.

File: src/main/java/com/atlauncher/mclauncher/MCLauncher.java
Comment (ellet0) [reply src/main/java/com/atlauncher/mclauncher/MCLauncher.java:448]: What about the single player, is there a way?

File: src/main/java/com/atlauncher/data/Instance.java
Comment (ellet0) [reply src/main/java/com/atlauncher/data/Instance.java]: Reverted.

File: src/main/java/com/atlauncher/mclauncher/MCLauncher.java
Comment (ellet0) [reply src/main/java/com/atlauncher/mclauncher/MCLauncher.java]: Done, doing testing now to make sure it work

Comment (ellet0) [reply src/main/java/com/atlauncher/mclauncher/MCLauncher.java]: > It might be good to do these checks in the settings page and disable the singleplayer and realms options in the dropdown if it's not supported so they cannot be selected.

Done, doing testing now to make sure it work

Comment (ellet0) [reply src/main/java/com/atlauncher/mclauncher/MCLauncher.java]: Updated in [325d2f42fe6e2ea8bef6979687a50720c01be3f4](https://github.com/ATLauncher/ATLauncher/pull/893/commits/325d2f42fe6e2ea8bef6979687a50720c01be3f4)

While I tested this feature with different cases, we might want to test it more before releasing, I can build the jar for you or you can build it yourself by cloning the fork and the branch for this feature, or you can approve it and then test it before release the new version

**Comment by ellet0** [comment]
@RyanTheAllmighty Can you re-run the workflow?

**Review by RyanTheAllmighty — APPROVED** [review]
Looks good, the last refactor makes it look much cleaner.

Thanks for this, been on the TODO list for a while

**Pull request merged**
